File: frontend/src/services/recipe.ts
/**
 * API service client for Recipe Management and Menu Item Ingredient Composition.
 */

const API = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

export interface RecipeIngredientInput {
  ingredient_id: string;
  quantity: number;
  unit_of_measure: "kg" | "g" | "L" | "ml" | "pcs";
}

export interface RecipeIngredientOut {
  id: string;
  ingredient_id: string;
  ingredient_name: string;
  quantity: number;
  unit_of_measure: string;
  current_stock: number;
  minimum_stock: number;
  unit_cost: number;
}

export interface RecipeCreatePayload {
  menu_item_id: string;
  name: string;
  instructions?: string;
  yields?: number;
  ingredients: RecipeIngredientInput[];
}

export interface RecipeOut {
  id: string;
  menu_item_id: string;
  menu_item_name?: string | null;
  name: string;
  instructions?: string | null;
  yields: number;
  ingredients: RecipeIngredientOut[];
  max_makeable_portions: number;
  is_available: boolean;
  created_at: string;
  updated_at: string;
}

async function recipeFetch(
  path: string,
  token: string,
  init?: RequestInit
): Promise<Response> {
  return fetch(`${API}/api/v1/recipes${path}`, {
    ...init,
    headers: {
      Authorization: `Bearer ${token}`,
      ...(init?.headers as Record<string, string> ?? {}),
    },
  });
}

export async function fetchAllRecipes(token: string): Promise<RecipeOut[]> {
  const res = await recipeFetch("", token);
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err?.detail ?? "Failed to fetch recipes.");
  }
  return res.json();
}

export async function fetchRecipeByMenuItemId(
  token: string,
  menuItemId: string
): Promise<RecipeOut | null> {
  const res = await recipeFetch(`/menu-item/${menuItemId}`, token);
  if (!res.ok) {
    if (res.status === 404) return null;
    const err = await res.json().catch(() => ({}));
    throw new Error(err?.detail ?? "Failed to fetch recipe.");
  }
  return res.json();
}

export async function saveRecipe(
  token: string,
  payload: RecipeCreatePayload
): Promise<RecipeOut> {
  const res = await recipeFetch("", token, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err?.detail ?? "Failed to save recipe.");
  }
  return res.json();
}

export async function deleteRecipe(
  token: string,
  recipeId: string
): Promise<void> {
  const res = await recipeFetch(`/${recipeId}`, token, {
    method: "DELETE",
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err?.detail ?? "Failed to delete recipe.");
  }
}

File: backend/app/schemas/recipe.py
"""
Pydantic schemas for Recipe Management, Ingredient Composition, and Stock Deductions.
"""

import uuid
from datetime import datetime
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class RecipeIngredientInput(BaseModel):
    ingredient_id: uuid.UUID
    quantity: float = Field(
        ...,
        gt=0.0,
        description="Required quantity per portion must be strictly positive",
    )
    unit_of_measure: Literal["kg", "g", "L", "ml", "pcs"] = Field(
        ..., description="Unit of measurement"
    )


class RecipeIngredientOut(BaseModel):
    id: uuid.UUID
    ingredient_id: uuid.UUID
    ingredient_name: str
    quantity: float
    unit_of_measure: str
    current_stock: float
    minimum_stock: float
    unit_cost: float

    model_config = ConfigDict(from_attributes=True)


class RecipeCreatePayload(BaseModel):
    menu_item_id: uuid.UUID
    name: Annotated[str, Field(min_length=1, max_length=255)]
    instructions: Annotated[str | None, Field(max_length=2000)] = None
    yields: float = Field(default=1.0, gt=0.0)
    ingredients: list[RecipeIngredientInput] = Field(..., min_length=1)

    @field_validator("ingredients")
    @classmethod
    def validate_no_duplicate_ingredients(
        cls, value: list[RecipeIngredientInput]
    ) -> list[RecipeIngredientInput]:
        seen = set()
        for item in value:
            if item.ingredient_id in seen:
                raise ValueError(
                    "Duplicate ingredients in a single recipe are not allowed."
                )
            seen.add(item.ingredient_id)
        return value


class RecipeOut(BaseModel):
    id: uuid.UUID
    menu_item_id: uuid.UUID
    menu_item_name: str | None = None
    name: str
    instructions: str | None = None
    yields: float
    ingredients: list[RecipeIngredientOut]
    max_makeable_portions: int
    is_available: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
